### PR TITLE
100-continue

### PR DESCRIPTION
**Pull request by jaschrep-msft**
Enables a strategy for adding `Expect: 100-continue` to requests when the service may be under load.

- On/off/auto modes for adding these headers to requests that container a content body.
  - On always adds the header when a content body reaches a length threshold.
  - Off never adds the header.
  - Auto will add the header when a content body reaches a length threshold, and the client has received a 429, 500, or 503 response within a recent time span.
- Default auto timespan is 1 minute.
- Default length threshold is 0.
- Default mode is auto

Future PR work:
- Compat switch for getting old behavior without code change
- Achieve this behavior on RST

### CODE REVIEW

**Review by seanmcc-msft — COMMENTED** [review]
Looks good to me!

Have we tried manually testing in "On" mode and Fiddler enabled as a sanity check?

I am also wondering whether or not we should update the changelogs for this feature.

File: sdk/storage/Azure.Storage.Common/src/Shared/ExpectContinueOnThrottlePolicy.cs [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+        }
+    }
+
+    public override void OnReceivedResponse(HttpMessage message)
+    {
+        base.OnReceivedResponse(message);
+        if (message.HasResponse && (
+            message.Response.Status == 429 ||
+            message.Response.Status == 500 ||
+            message.Response.Status == 503))
```
Comment (jalauzon-msft) [inline sdk/storage/Azure.Storage.Common/src/Shared/ExpectContinueOnThrottlePolicy.cs:42]: I think we also wanted to enable 100-continue on TCP RST? According to the email thread this would be `SocketException` with a [SocketError](https://learn.microsoft.com/en-us/dotnet/api/system.net.sockets.socketerror?view=net-8.0) of `ConnectionReset`. Not sure how that would manifest in the pipeline message though.

Email also mentions "timeouts" but I'm a bit unclear on what type of timeout they mean. Likely network timeouts and/or HTTP 408?

Comment (jaschrep-msft) [reply sdk/storage/Azure.Storage.Common/src/Shared/ExpectContinueOnThrottlePolicy.cs:42]: Yes I plan to add support for RST in a future PR. The main issue is that RST doesn't just happen in the pipeline, and so it can't be handled just there. While this policy needs to be what adds the header, we need a broader solution to have failures like that from elsewhere in the library signal the error has been encountered.

**Comment by jaschrep-msft** [comment]
> Have we tried manually testing in "On" mode and Fiddler enabled as a sanity check?

@seanmcc-msft I can add a test into the blobs package that checks the headers as they leave the pipeline to go to the transport layer.

File: sdk/storage/Azure.Storage.Blobs/api/Azure.Storage.Blobs.net6.0.cs [hunk 2]
```diff
@@ -56,6 +56,7 @@ public partial class BlobClientOptions : Azure.Core.ClientOptions
         public Azure.Storage.Blobs.Models.CustomerProvidedKey? CustomerProvidedKey { get { throw null; } set { } }
         public bool EnableTenantDiscovery { get { throw null; } set { } }
         public string EncryptionScope { get { throw null; } set { } }
+        public Azure.Storage.ExpectContinueOptions ExpectContinueBehavior { get { throw null; } set { } }
```
Comment (tg-msft) [inline sdk/storage/Azure.Storage.Blobs/api/Azure.Storage.Blobs.net6.0.cs:59]: Do regular developers know what `ExpectContinue` is?  This feels more like the "how" not the "why."  Should we come up with a more intuitive name like `(Cautious|Preflight|...)(Upload|Transfer|...)`?  Ideally a better name would also let us follow our usual pattern of `public FooOptions Foo { get; set; }` for *Options types in *ClientOptions.

Does this need to be top-level or should it be nested in `StorageTransferOptions`?

File: sdk/storage/Azure.Storage.Common/api/Azure.Storage.Common.net6.0.cs [hunk 3]
```diff
@@ -20,6 +20,19 @@ public partial class DownloadTransferValidationOptions
         public bool AutoValidateChecksum { get { throw null; } set { } }
         public Azure.Storage.StorageChecksumAlgorithm ChecksumAlgorithm { get { throw null; } set { } }
     }
+    public partial class ExpectContinueOptions
+    {
+        public ExpectContinueOptions() { }
+        public System.TimeSpan Backoff { get { throw null; } set { } }
```
Comment (tg-msft) [inline sdk/storage/Azure.Storage.Common/api/Azure.Storage.Common.net6.0.cs:26]: This name looks more like it's for a retry behavior than something that configures how long `Auto` stays on for.  Consider a name more like like `AutoEnabledInterval`?

File: sdk/storage/Azure.Storage.Common/api/Azure.Storage.Common.net6.0.cs [hunk 4]
```diff
@@ -20,6 +20,19 @@ public partial class DownloadTransferValidationOptions
...
+        public long ContentLengthThreshold { get { throw null; } set { } }
```
Comment (tg-msft) [inline sdk/storage/Azure.Storage.Common/api/Azure.Storage.Common.net6.0.cs:27]: This should probably be nullable if we're not requiring it.  Is this only for Auto?  Is it a min or a max?  Can we make the name more specific?

File: sdk/storage/Azure.Storage.Common/api/Azure.Storage.Common.net6.0.cs [hunk 5]
```diff
@@ -20,6 +20,19 @@ public partial class DownloadTransferValidationOptions
...
+        public Azure.Storage.ExpectContinueOptions.ApplyHeaderMode Mode { get { throw null; } set { } }
+        public enum ApplyHeaderMode
```
Comment (tg-msft) [inline sdk/storage/Azure.Storage.Common/api/Azure.Storage.Common.net6.0.cs:29]: Nested types are confusing for our average customer and the intellisense driven experience in C#.  The name should also be more about the user scenario enabled rather than the implementation.

File: sdk/storage/Azure.Storage.Common/api/Azure.Storage.Common.net6.0.cs [hunk 6]
```diff
@@ -20,6 +20,19 @@ public partial class DownloadTransferValidationOptions
...
+        {
+            Auto = 0,
```
Comment (tg-msft) [inline sdk/storage/Azure.Storage.Common/api/Azure.Storage.Common.net6.0.cs:31]: Can we make this name more descriptive to imply it's for throttling so people will better understand why it's important?

File: sdk/storage/Azure.Storage.Common/api/Azure.Storage.Common.net6.0.cs [hunk 7]
```diff
@@ -20,6 +20,19 @@ public partial class DownloadTransferValidationOptions
...
+            On = 1,
+            Off = 2,
```
Comment (tg-msft) [inline sdk/storage/Azure.Storage.Common/api/Azure.Storage.Common.net6.0.cs:33]: `Off` should probably be `=0`.

**Review by tg-msft — COMMENTED** [review]
Adding a couple of quick comments about the API and I'll try to review the implementation a little later too.

File: sdk/storage/Azure.Storage.Blobs/api/Azure.Storage.Blobs.net6.0.cs (hunk 2 above)
Comment (jaschrep-msft) [reply sdk/storage/Azure.Storage.Blobs/api/Azure.Storage.Blobs.net6.0.cs:59]: I'm not really sure what a good name would be, given it's a very specific header with understood behavior. Even if we come up with a better top level name, I would still expect we use this language for the individual properties of the class.

I don't think this can go in `StorageTransferOptions`. That is only available on a subset of our transfers, whereas we want this on any request with a content body.

File: sdk/storage/Azure.Storage.Common/api/Azure.Storage.Common.net6.0.cs (hunk 5 above)
Comment (jaschrep-msft) [reply sdk/storage/Azure.Storage.Common/api/Azure.Storage.Common.net6.0.cs:29]: See my previous comment that this is a very specific, well-known http header that has a specific behavior. Yes it's an implementation detail but I'm really not sure it's one we want to hide. Especially since we expect most users to go along with what we set and only advanced ones may want to go in and change it.

File: sdk/storage/Azure.Storage.Common/api/Azure.Storage.Common.net6.0.cs (hunk 7 above)
Comment (jaschrep-msft) [reply sdk/storage/Azure.Storage.Common/api/Azure.Storage.Common.net6.0.cs:33]: Disagree. 0 should be the default behavior as I understand it, which this is.

File: sdk/storage/Azure.Storage.Common/api/Azure.Storage.Common.net6.0.cs (hunk 4 above)
Comment (jaschrep-msft) [reply sdk/storage/Azure.Storage.Common/api/Azure.Storage.Common.net6.0.cs:27]: This is for both currently. And yes this can be nullable. I'm not sure how to be more descriptive for this name though without becoming word soup. `ContentLengthApplyExpectHeaderThreshold` starts to get pretty long while we do have a good docstring for the property. I think this comes back to having a snappy name for the options as a whole rather than using "expect continue." If that gets solved, I can probably think of a better name here.

File: sdk/storage/Azure.Storage.Common/api/Azure.Storage.Common.net6.0.cs (hunk 5 above)
Comment (jaschrep-msft) [reply sdk/storage/Azure.Storage.Common/api/Azure.Storage.Common.net6.0.cs:29]: Currently have extracted the enum to separate file and named it `ExpectContinueOptions`. Still think we need to use a name similar to "expect continue" somewhere in this API.

File: sdk/storage/Azure.Storage.Common/src/Shared/StorageClientOptions.cs [hunk 8]
```diff
@@ -122,6 +126,33 @@ public static HttpPipeline Build(this ClientOptions options, HttpPipelinePolicy
...
+                            ContentLengthThreshold = expectContinue.ContentLengthThreshold ?? 0,
+                        });
+                        break;
+                    case ExpectContinueMode.Off:
+                        break;
+                }
+            }
+            else
+            {
+                // TODO get env config for whether to disable
```
Comment (amnguye) [inline sdk/storage/Azure.Storage.Common/src/Shared/StorageClientOptions.cs:152]: Is this going to be done in a separate PR? Or should we wait on the ask for this

**Review by amnguye — APPROVED** [review]

File: sdk/storage/Azure.Storage.Common/src/Shared/StorageClientOptions.cs (hunk 8 above)
Comment (jaschrep-msft) [reply sdk/storage/Azure.Storage.Common/src/Shared/StorageClientOptions.cs:152]: Yes, this is future PR work. This will be necessary before release but I'm avoiding bloating the PR.
